pkg/intermediate/aggregate.go

```diff
@@ -124,74 +140,272 @@ func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
 					}
+					existingIeWithValue.Value = ieWithValue.Value
 				}
+			case entities.Unsigned16:
+				if ieWithValue.Value != uint16(0) {
+					existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
+					if existingIeWithValue.Value != uint16(0) {
+						klog.Warningf("This field with name %v should not have been filled with value %v in existing record.", field, existingIeWithValue.Value)
+					}
+					existingIeWithValue.Value = ieWithValue.Value
+				}
+			case entities.Ipv4Address:
+				ipInString := ieWithValue.Value.(net.IP).To4().String()
+				if ipInString != "0.0.0.0" {
+					existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
+					ipInString := existingIeWithValue.Value.(net.IP).To4().String()
+					if ipInString != "0.0.0.0" {
+						klog.Warningf("This field with name %v should not have been filled with value %v in existing record.", field, existingIeWithValue.Value)
+					}
+					existingIeWithValue.Value = ieWithValue.Value
+				}
+			case entities.Ipv6Address:
+				ipInString := ieWithValue.Value.(net.IP).To16().String()
+				if ipInString != net.ParseIP("::0").To16().String() {
+					existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
+					ipInString := existingIeWithValue.Value.(net.IP).To16().String()
+					if ipInString != net.ParseIP("::0").To16().String() {
+						klog.Warningf("This field with name %v should not have been filled with value %v in existing record.", field, existingIeWithValue.Value)
+					}
+					existingIeWithValue.Value = ieWithValue.Value
+				}
+			default:
+				klog.Errorf("Fields with dataType %v is not supported in correlation fields list.", ieWithValue.Element.DataType)
 			}
 		}
 	}
-	a.addRecordToMap(flowKey, record)
-	a.removeDuplicates(flowKey)
 }
 
-// removeDuplicates is currently used only in correlateRecords().
-// For other uses, please acquire the flowKeyRecordLock for protection.
-func (a *AggregationProcess) removeDuplicates(flowKey FlowKey) {
-	records := a.flowKeyRecordMap[flowKey]
-	srcRecords := make([]entities.Record, 0)
-	dstRecords := make([]entities.Record, 0)
-	for _, record := range records {
-		if isRecordFromSrc(record) {
-			srcRecords = append(srcRecords, record)
-		} else {
-			dstRecords = append(dstRecords, record)
+// aggregateRecords aggregate the incomingRecord with existingRecord by updating
+// stats and flow timestamps.
+func (a *AggregationProcess) aggregateRecords(incomingRecord, existingRecord entities.Record, fillSrcStats, fillDstStats bool) {
+	if a.aggregateElements == nil {
+		return
+	}
+
+	for _, field := range a.aggregateElements.nonStatsElements {
+		if ieWithValue, exist := incomingRecord.GetInfoElementWithValue(field); exist {
+			switch ieWithValue.Element.Name {
+			case "flowEndSeconds":
+				existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
```

Comment:
So no need to check if the field exists in existing record?

Reply:
Yes not needed as we do the check, we already check for this in addFieldsForStatsAggregation. The assumption is that all the flow records for a given flow follow the same template.